File: modules/expo-driver-location/ios/DriverLocLog.swift
import Foundation

/**
 * Single console tag for the whole driver-location system.
 * Filter in Xcode / Console.app: DriverLoc
 */
enum DriverLocLog {
  private static let tag = "DriverLoc"

  static func i(_ event: String, _ details: String = "") {
    print("[\(tag)] \(format(event, details))")
  }

  static func w(_ event: String, _ details: String = "") {
    print("[\(tag)] \(format(event, details))")
  }

  static func e(_ event: String, _ details: String = "") {
    print("[\(tag)] \(format(event, details))")
  }

  private static func format(_ event: String, _ details: String) -> String {
    if details.isEmpty {
      return "event=\(event)"
    }
    return "event=\(event) | \(details)"
  }

  static func coord(
    lat: Double?,
    lon: Double?,
    accuracy: Double? = nil,
    capturedAtMs: Double? = nil
  ) -> String {
    var parts = [
      "lat=\(lat.map { String($0) } ?? "-")",
      "lon=\(lon.map { String($0) } ?? "-")",
    ]
    if let accuracy {
      parts.append("accuracy=\(accuracy)")
    }
    if let capturedAtMs, capturedAtMs > 0 {
      let age = max(0, Int((Date().timeIntervalSince1970 * 1000) - capturedAtMs))
      parts.append("capturedAt=\(Int(capturedAtMs))")
      parts.append("ageMs=\(age)")
    }
    return parts.joined(separator: " ")
  }
}

File: modules/expo-driver-location/android/src/main/java/expo/modules/driverlocation/DriverLocLog.kt
package expo.modules.driverlocation

import android.util.Log

/**
 * Single Logcat tag for the whole driver-location system.
 * Filter after app kill: `adb logcat -s DriverLoc`
 */
object DriverLocLog {
  const val TAG = "DriverLoc"

  fun i(event: String, details: String = "") {
    Log.i(TAG, format(event, details))
  }

  fun w(event: String, details: String = "") {
    Log.w(TAG, format(event, details))
  }

  fun e(event: String, details: String = "", throwable: Throwable? = null) {
    if (throwable != null) {
      Log.e(TAG, format(event, details), throwable)
    } else {
      Log.e(TAG, format(event, details))
    }
  }

  private fun format(event: String, details: String): String {
    return if (details.isBlank()) {
      "event=$event"
    } else {
      "event=$event | $details"
    }
  }

  fun coord(
    lat: Double?,
    lon: Double?,
    accuracy: Double? = null,
    capturedAtMs: Double? = null,
  ): String {
    val parts = mutableListOf(
      "lat=${lat ?: "-"}",
      "lon=${lon ?: "-"}",
    )
    if (accuracy != null) {
      parts.add("accuracy=$accuracy")
    }
    if (capturedAtMs != null && capturedAtMs > 0) {
      parts.add("capturedAt=${capturedAtMs.toLong()}")
      parts.add("ageMs=${(System.currentTimeMillis() - capturedAtMs.toLong()).coerceAtLeast(0)}")
    }
    return parts.joinToString(" ")
  }
}
